File: infinity-server/src/main/java/io/github/derechtepilz/infinity/gamemode/serializer/ExperienceSerializer.java
package io.github.derechtepilz.infinity.gamemode.serializer;

import io.github.derechtepilz.infinity.Infinity;
import org.bukkit.entity.Player;
import org.bukkit.util.io.BukkitObjectInputStream;
import org.bukkit.util.io.BukkitObjectOutputStream;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;

public class ExperienceSerializer {

	private ExperienceSerializer() {}

	public static String serialize(Player player) {
		try {
			ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
			BukkitObjectOutputStream bukkitOutputStream = new BukkitObjectOutputStream(outputStream);
			bukkitOutputStream.writeInt(player.getLevel());
			bukkitOutputStream.writeFloat(player.getExp());
			bukkitOutputStream.close();
			return Base64.getEncoder().encodeToString(outputStream.toByteArray());
		} catch (IOException e) {
			Infinity.getInstance().getLogger().severe("There was an error while serializing experience. This might be a bug. Please report this!");
			for (StackTraceElement stackTraceElement : e.getStackTrace()) {
				Infinity.getInstance().getLogger().severe(stackTraceElement.toString());
			}
			return null;
		}
	}

	public static List<? extends Number> deserialize(String data) {
		try {
			ByteArrayInputStream inputStream = new ByteArrayInputStream(Base64.getDecoder().decode(data));
			BukkitObjectInputStream bukkitInputStream = new BukkitObjectInputStream(inputStream);
			int level = bukkitInputStream.readInt();
			float progress = bukkitInputStream.readFloat();
			bukkitInputStream.close();
			return new ArrayList<>(List.of(level, progress));
		} catch (IOException e) {
			Infinity.getInstance().getLogger().severe("There was an error while deserializing experience. This might be a bug. Please report this!");
			for (StackTraceElement stackTraceElement : e.getStackTrace()) {
				Infinity.getInstance().getLogger().severe(stackTraceElement.toString());
			}
			return null;
		}
	}

}

File: src/main/kotlin/io/github/derechtepilz/infinity/gamemode/experience/ExperienceSerializer.kt
package io.github.derechtepilz.infinity.gamemode.experience

import com.google.common.base.Preconditions
import org.bukkit.entity.Player
import org.bukkit.util.io.BukkitObjectInputStream
import org.bukkit.util.io.BukkitObjectOutputStream
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectOutputStream
import java.util.Base64
import kotlin.jvm.internal.Intrinsics

object ExperienceSerializer {

	@JvmStatic
	fun serializeLevel(level: Int): String {
		val outputStream = ByteArrayOutputStream()
		val bukkitOutputStream = BukkitObjectOutputStream(outputStream)
		bukkitOutputStream.writeInt(level)
		bukkitOutputStream.close()
		return Base64.getEncoder().encodeToString(outputStream.toByteArray())
	}

	@JvmStatic
	fun serializeProgress(progress: Float): String {
		val outputStream = ByteArrayOutputStream()
		val bukkitOutputStream = BukkitObjectOutputStream(outputStream)
		bukkitOutputStream.writeFloat(progress)
		bukkitOutputStream.close()
		return Base64.getEncoder().encodeToString(outputStream.toByteArray())
	}

	@JvmStatic
	fun deserialize(data: String, dataClass: Class<*>): Any {
		Preconditions.checkArgument(dataClass.simpleName == "int" || dataClass.simpleName == "float")
		val inputStream = ByteArrayInputStream(Base64.getDecoder().decode(data))
		val bukkitInputStream = BukkitObjectInputStream(inputStream)
		val experienceData = if (dataClass.simpleName == "int") bukkitInputStream.readInt() else bukkitInputStream.readFloat()
		bukkitInputStream.close()
		return experienceData
	}

}